tools/database/dbc/item_effect.go

```diff
@@ -33,3 +41,210 @@ func (e *ItemEffect) ToMap() map[string]interface{} {
+	for _, se := range dbcInstance.SpellEffects[spellID] {
+		if se.EffectAura == A_PROC_TRIGGER_SPELL {
+			return resolveStatsSpell(se.EffectTriggerSpell)
+		}
+	}
+	return spellID
+}
+
+func resolveTriggerType(topType, spellID int) int {
+	if topType == ITEM_SPELLTRIGGER_ON_USE || topType == ITEM_SPELLTRIGGER_CHANCE_ON_HIT {
+		return topType
+	}
+	for _, se := range dbcInstance.SpellEffects[spellID] {
+		if se.EffectAura == A_PROC_TRIGGER_SPELL {
+			return resolveTriggerType(ITEM_SPELLTRIGGER_CHANCE_ON_HIT, se.EffectTriggerSpell)
+		}
+	}
+	return topType
+}
+
+func buildScalingProps(spellID, itemLevel int) *proto.ScalingItemEffectProperties {
+	total := collectStats(spellID, itemLevel)
+	return &proto.ScalingItemEffectProperties{Stats: total.ToProtoMap()}
+}
+
+func collectStats(spellID, itemLevel int) stats.Stats {
+	var total stats.Stats
+
+	var emptyStats = stats.Stats{}
+	visited := make(map[int]bool)
+
+	var recurse func(int)
+	recurse = func(id int) {
+		if visited[id] {
+			return
+		}
+		visited[id] = true
+
+		sp := dbcInstance.Spells[id]
+		for _, se := range dbcInstance.SpellEffects[id] {
+			if s := se.ParseStatEffect(sp.HasAttributeAt(11, 0x4), itemLevel); s != &emptyStats {
+				total.AddInplace(s)
+			} else if se.EffectAura == A_PROC_TRIGGER_SPELL {
+				recurse(se.EffectTriggerSpell)
+			}
+		}
+	}
+
+	recurse(spellID)
+	return total
+}
+
+func ParseItemEffects(itemID, itemLevel int, levelState proto.ItemLevelState) []*proto.ItemEffect {
+	raw := dbcInstance.ItemEffectsByParentID[itemID]
+	out := make([]*proto.ItemEffect, 0, len(raw))
+	for _, ie := range raw {
+		if pe, ok := ie.ToProto(itemLevel, levelState); ok {
+			out = append(out, pe)
+		}
+	}
+	return out
+}
+
+// Parses a UIItem and loops through Scaling Options for that item.
+func MergeItemEffectsForAllStatesNew(parsed *proto.UIItem) *proto.ItemEffect {
+	// pick a base effect that has stats if there is more than one effect on the item
+	var baseEff *ItemEffect
+	for i := range dbcInstance.ItemEffectsByParentID[int(parsed.Id)] {
+
+		e := &dbcInstance.ItemEffectsByParentID[int(parsed.Id)][i]
+		props := buildScalingProps(resolveStatsSpell(e.SpellID), int(parsed.ScalingOptions[int32(proto.ItemLevelState_Base)].Ilvl))
+		if len(props.Stats) > 0 {
+			baseEff = e
+			break
+		}
+	}
+	if baseEff == nil {
+		return nil
+	}
+	statsSpellID := resolveStatsSpell(baseEff.SpellID)
+	pe := makeBaseProto(baseEff, statsSpellID)
+	assignTrigger(baseEff, statsSpellID, pe)
+
+	// add scaling for each saved state
+	for state, opt := range parsed.ScalingOptions {
+		if parsed.Id == 92127 {
+			fmt.Println("hello2", parsed, parsed.Ilvl, opt.Ilvl, opt)
+		}
+		ilvl := int(opt.Ilvl)
+		props := buildScalingProps(statsSpellID, ilvl)
+		pe.ScalingOptions[state] = props
+
+		if proc := pe.GetProc(); proc != nil {
+			_, specMods := realPpmModifier(dbcInstance.Spells[statsSpellID], ilvl)
+			proc.SpecModifiers = specMods
+		}
+	}
+
+	return pe
+}
+
+func realPpmScale(spell Spell) int {
```

Comment:
You probably want to take in a `*Spell` pointer here instead, in order to avoid copying the input on every call.